src/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,653 @@
+
+    /// @dev convention for all double nested address mappings is key (venue -> user) -> value
+
+    /// @notice record how much VCON a user deposited in a given venue
+    mapping(address => mapping(address => uint256)) public venueUserShares;
+
+    /// @param _core reference to core
+    /// @param _pcvRouter reference to pcvRouter
+    constructor(address _core, address _pcvRouter) CoreRefV2(_core) {
+        pcvRouter = _pcvRouter;
+    }
+
+    /// @notice update the VCON share price and the last recorded profit for a given venue
+    /// @param venue address to accrue
+    function accrueVcon(address venue) external globalLock(1) whenNotPaused {
+        IPCVOracle oracle = pcvOracle();
+
+        require(oracle.isVenue(venue), "MarketGovernance: invalid destination");
+
+        _accrue(venue);
+    }
+
+    /// ---------- Permissionless User PCV Allocation Methods ----------
+
+    /// @notice a user can get slashed up to their full VCON stake for entering
+    /// a venue that takes a loss.
+    /// @param amountVcon to stake on destination
+    /// @param destination address to accrue rewards to, and send funds to
+    function stake(
+        uint256 amountVcon,
+        address destination
+    ) external globalLock(1) whenNotPaused {
+        IPCVOracle oracle = pcvOracle();
+        require(
+            oracle.isVenue(destination),
+            "MarketGovernance: invalid destination"
+        );
+
+        _accrue(destination); /// update share price in the destination so the user gets in at the current share price
+
+        /// vconToShares will always return correctly as accrue will set venueLastRecordedVconSharePrice
+        /// to the correct share price from 0 if uninitialized
+        uint256 userShareAmount = vconToShares(destination, amountVcon);
+
+        /// user updates
+        venueUserShares[destination][msg.sender] += userShareAmount;
+
+        /// venue updates
+        venueTotalShares[destination] += userShareAmount;
+
+        /// check and an interaction with a trusted contract
+        vcon().safeTransferFrom(msg.sender, address(this), amountVcon); /// transfer VCON in
+
+        emit Staked(destination, msg.sender, amountVcon);
+    }
+
+    /// @notice unstake VCON and transfer corresponding VCON to another venue
+    /// @param shareAmount the amount of shares to unstake
+    /// @param source address to accrue rewards to, and pull funds from
+    /// @param vconRecipient address to receive the VCON
+    /// @dev both source and destination are checked twice,
+    /// the first time in the market governance contract,
+    /// the second time in the PCV Router contract.
+    function unstake(
+        uint256 shareAmount,
+        address source,
+        address vconRecipient
+    ) external globalLock(1) whenNotPaused {
+        /// ---------- Checks ----------
+        IPCVOracle oracle = pcvOracle();
+
+        require(oracle.isVenue(source), "MarketGovernance: invalid source");
```

Comment:
you don't need to do this, because `PCVRouter.movePCV()` will check, and also the mapping below will return 0

Reply:
It's really bad practice to allow un sanitized inputs to get all the way to the PCVRouter. The market governance contract needs to verify before calling the PCV Router that all inputs are correct, even if the PCV Router double checks things again.